lib/connect.js

```diff
@@ -234,25 +240,86 @@ const getTasks = (model, setupListeners) => {
         setupListeners(mongoClient);
       }
 
+      let maybeTunnelError;
+
+      if (tunnel) {
+        tunnel.on('error', (tunnelError) => {
+          if (tunnelError.origin === 'ssh-client') {
+            maybeTunnelError = tunnelError;
+          }
+        });
+      }
+
       mongoClient.connect((err, _client) => {
+        // When mongo client failed to connect AND tunnel encountered an error
+        // those two are totally related and we want to return a tunnel error as
+        // it will have more relevant context for the end user
+        err = err && maybeTunnelError ? maybeTunnelError : err;
+
         ctx(err);
 
         if (err) {
           if (tunnel) {
             debug('data-service connection error, shutting down ssh tunnel');
-            tunnel.close();
+            tunnel.close().then(
+              () => {
+                debug('ssh tunnel stopped');
+                cb(err);
+              },
+              (tunnelError) => {
+                debug('ssh tunnel stopped with error: %s', tunnelError.message);
+                cb(err);
+              }
+            );
           }
-
-          return cb(err);
+          return;
         }
 
         client = _client;
 
         if (tunnel) {
-          client.on('close', () => {
-            debug('data-service disconnected. shutting down ssh tunnel');
-            tunnel.close();
-          });
+          const origClose = client.close;
+
+          client.close = function close(force, callback) {
+            if (typeof force === 'function') {
+              callback = force;
+              force = false;
+            }
+
+            const promise = new Promise((res, rej) => {
+              function callbackOrPromise(error = null) {
+                if (callback) {
+                  return callback(error);
+                }
+
+                if (error) {
+                  return rej(error);
+                }
+
+                res();
+              }
+
+              origClose.call(this, force, (clientError) => {
+                debug('data-service disconnected. shutting down ssh tunnel');
+
+                tunnel.close().then(
+                  () => {
+                    debug('ssh tunnel stopped');
+                    callbackOrPromise(clientError);
```

Comment:
why we return an error here? Isn't this what is supposed to happen?

Reply:
This case is for when client failed to connect. We wait for tunnel to close and only then resolve the callback with the original client error